src/dstore/dstore_base.c

```diff
@@ -799,3 +757,125 @@ int dstore_pread(struct dstore_obj *obj, off_t offset, size_t count,
 
 	return rc;
 }
+
+int dstore_dealloc_op(struct dstore_obj *obj, struct dstore_io_vec *vec,
+		      struct dstore_io_op **out)
+{
+	int rc;
+
+	rc = dstore_io_op_init_and_submit(obj, vec, out,
+					  DSTORE_IO_OP_FREE);
+
+	log_debug("dstore_dealloc_op (" OBJ_ID_F " <=> %p, "
+		  "vec=%p *out=%p) rc=%d",
+		  OBJ_ID_P(dstore_obj_id(obj)), obj, vec,
+		  rc == 0 ? *out : NULL, rc);
+
+	return rc;
+}
+
+int dstore_deallocate(struct dstore_obj *obj, off_t offset, size_t count)
+{
+	int rc = 0;
+
+	dassert(obj);
+
+	char *tmp_buf = NULL;
+	struct dstore_io_op *wop = NULL;
+	struct dstore_io_vec *vec = NULL;
+	size_t nr_request;
+	size_t tail_size;
+	size_t nblk_per_req;
+	size_t ndata_per_req;
+	int i;
+
+	size_t bsize = dstore_get_bsize(obj->ds,
+					(dstore_oid_t *)dstore_obj_id(obj));
+
+	if (offset % bsize != 0) {
+		/*we need to make deallocate operation left aligned */
+		tmp_buf = calloc(bsize, sizeof(char));
+		uint32_t left_blk_num = offset/bsize;
+		uint32_t write_count = offset - (left_blk_num * bsize);
+		write_count = bsize - write_count;
+
+		if (count < write_count) {
+
+			write_count = count;
+		}
+
+		RC_WRAP_LABEL(rc, out, dstore_pwrite, obj,
+			      offset, write_count,
+			      bsize, tmp_buf);
+
+		if (write_count == count) {
+			goto out;
+		}
+
+		count = count - write_count;
+		offset = offset + write_count;
+	}
+
+	ndata_per_req = DSAL_MAX_IO_SIZE;	
+	nr_request = count / ndata_per_req;
+	tail_size = count - (nr_request * ndata_per_req);
+	nblk_per_req = ndata_per_req / bsize;
+
+	vec = calloc(1, sizeof(struct dstore_io_vec));
+        if (vec == NULL) {
+                rc = -ENOMEM;
+                goto out;
+        }
+	/* we do not need io buffer here, we just need to send extents info */
+	vec->flags |= IO_NO_DATA;
+	dstore_io_vec_set_from_edbuf(vec);
+
+	for (i = 0; i < nr_request; i++) {
+
+		*(vec->ovec) = offset;
+		*(vec->svec) = ndata_per_req;
+
+		RC_WRAP_LABEL(rc, out, dstore_dealloc_op, obj, vec, &wop);
+
+		RC_WRAP_LABEL(rc, out, dstore_io_op_wait, wop);
+
+		dstore_io_op_fini(wop);
+
+		wop = NULL;
+
+		offset += ndata_per_req;
+	}
+
+	if (tail_size) {
+
```

Comment:
Question: @ivan-alekhin  and @gshipra , Do we really need this right-most unaligned part? I believe we will never be un-aligned in right-most part because while writing this data we would have taken care and written always in aligned boundary.

NOTE: In old code also we were not doing `MC_OC_FREE` for right most part.

Reply:
that's a good question Anil. 

In my original code the idea was to round up the number of blocks, so that the right part (if is not aligned) will be de-allocated:

```c
	/* ajust to the very first byte of the nearest (from right) page */
	aligned_off = m0_round_up(offset, bsize);
	/* cut out the left unaligned part from the whole len,
	 * round it up and then count the amount of blocks */
	nblocks = m0_round_up((count + offset) - aligned_off, bsize) / bsize;
// ...
	rc = m0_file_unmap_aligned(fid, nblocks, aligned_off, bsize);

// and then inside unmap_aligned:
	ndata_per_req = lower(trunc_data_per_request, bsize);
	nblk_per_req = ndata_per_req / bsize;
	nrequests = nblocks / nblk_per_req;
	tail_size = (nblocks * bsize) - (nrequests * ndata_per_req);
```

So that, `tail_size` is juts the aligned leftover from the right side. It should be aligned with the block size but should not be aligned with the request size. The following invariant should be held for it: `tail_size % block_size == 0 && tail_size < ndata_per_req`.

To conclude, I guess we are handling here unaligned right side here in a wrong way: if `offset+count % bs != 0`  then tail_size also won't be aligned (`tail_size % bs != 0`). It means M0 client will tell us that we using the wrong input.

@gshipra , what do you think about moving the portion of code that handles aligned FREE operation into its own function (as it was in the original code)? it will help us to analyze the code.

Reply:
@Anilsindhi  - This is not right alignment for data buffer. We are dividing FREE in multiple of 20 MB. Lets sat we have 21 MB to free. Tail size here is this 21%20 = 1 MB.

Reply:
@Anilsindhi  - Consider follwoing scenarios.

3000 bytes are written to a file at offset 0, we will round up write request and actually write 4096 byes. File Size is 3000.

1. Trunc request to truncate file to 0. Now offset is 0 and we need to free 3000 bytes.
We will send count as 3000 to free operation as per current implementation. But motr does not handle this. It panics.
Motr needs aligned count. I figured this out while testing. 
When I tried to free 4096 bytes, operation succeeded.

2. Lets say truncate sets the new file size to 9000. As per current implementation we wont do any thing other than increasing the size.
0-3000 - valid data
3000 - 4095 - 0s
4096-8191 - hole
8192 - 9000 - ?????

Reply:
@Anilsindhi , @ivan-alekhin - please suggest which function is better to do left+right alignment calculataions ? dstore_shrink or dstore_deallocate ?

Reply:
Okay I understood the use case. But just the question on your reply.
Your offset = 0, count = 3000 so,
```
if (offset % bsize != 0) {
// We won't go here as offset is zero so offset % bsize == 0
}

ndata_per_req = 20MB;
nr_request = count / ndata_per_req = 3000/20MB = 0;
tail_size = count - (nr_request * ndata_per_req) = 3000 - (0 *20MB) = 3000; // Not aligned to bsize

for (i = 0; i < nr_request; i++) {
      // We will not come here as `nr_request` is 0
}

// In the following code you are still issuing unaligned unmap request to motr.

	if (tail_size) {
		*(vec.ovec) = offset;
		*(vec.svec) = tail_size; // This is 3000 and not rounded up to bsize

		RC_WRAP_LABEL(rc, out, dstore_dealloc_op, obj, &vec, &dop);

		RC_WRAP_LABEL(rc, out, dstore_io_op_wait, dop);

		dstore_io_op_fini(dop);

		dop = NULL;
	}
```
If my understanding is correct then still the current code is not behaving as expected.

Reply:
@gshipra ,

> please suggest which function is better to do left+right alignment calculataions ? dstore_shrink or dstore_deallocate ?

dstore_shrink uses dstore_pwrite. dstore_pwrite is able to handle unaligned ranges for WRITE operation. But there are no function that can handle unaligned ranges for FREE operation. In fact, dstore_deallocate will be the function that will handle this. Therefore, you cannot use dstore_shrink as a reference. So, I'd recommend to look at the original code (
m0_file_unmap) -- there are two separate functions. One function handles any aligned range (by grouping blocks into bigger requests), and the other handles alignment (by writing zeros on the left side and freeing full blocks on the right side).

Reply:
@Anilsindhi - I have told in previous comments that two scenarios are discovered via new UT code added and are not handled in code yet. You mentioned in a comment before that we do not need to do right alignment but it seems we need to do right alignment else motr may panic for last free request.

Reply:
@gshipra, Okay I understood now that you are aware of the issue which I highlighted, good that we have written UT along with this change and we caught the issue. Please go through @ivan-alekhin comments in this thread, and look at the old code as per his suggestion and you can something out of that.

Reply:
Added the right alignment code.

Reply:
I am fine with this changes, @ivan-alekhin if you are also fine then feel free to close this comment.

Reply:
Yes, the recent changes look good.